src/Stratis.Bitcoin/Utilities/MemoryCache.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    public class MemoryCache<TKey, TValue>
+    {
+        /// <summary>Dictionary that contains cached items.</summary>
+        private readonly ConcurrentDictionary<TKey, TValue> cache;
+
+        /// <summary>Keys sorted by time they when they were accessed last time with most recent ones in the end.</summary>
+        private readonly LinkedList<TKey> keys;
+
+        /// <summary>Maximum items count that can be stored in the cache.</summary>
+        private readonly int sizeLimit;
+
+        /// <summary>Amount to compact the cache by when the maximum size is exceeded.</summary>
```

Comment:
if it is percentage then it is not amount
also naming something Percentage might invoke one to think that a value of e.g. `10` would mean 10 %
when you have a variable that has such strict bounds, it is always a good idea to explicitly document the bounds
i'd say
`Percentage of the items in the cache to be removed when the maximum item count is reached. This has to be a number within (0; 1] interval.`

and it very much makes sense to assert this boundary in the constructor as well

Reply:
Fixed by removing the Compact()